### PR TITLE
Feature 144298 layer extent

### PR DESCRIPTION
**Pull request by almostcake**
getBounding box method usage for wms and wfs layers, added fitBounds button to maplayers tree

### CODE REVIEW

File: addon/components/layers-dialogs/settings/wfs.js [hunk 1]
```diff
@@ -4,46 +4,128 @@
...
+
+  /**
+    OBSOLETE
+    Observes changes in settings.url property and checks whether it is correct url.
+    If url syntax is correct, requests available info formats from service.
+
+    @method _urlDidChange
+    @private
   */
-  tagName: '',
```
Comment (kuzkok) [inline addon/components/layers-dialogs/settings/wfs.js]: do you need div tag in this component?

File: addon/layers/-private/base.js [hunk 2]
```diff
@@ -79,7 +77,19 @@ export default Ember.Object.extend({
   */
   createSettings() {
     let settings = {
-      opacity: 1
+      opacity: 1,
+      bounds: [
+        [-90, -180],
+        [90, 180]
+      ],
+      wgs84bbox: [
```
Comment (kuzkok) [inline addon/layers/-private/base.js]: why not wgs84bbox: null?

File: addon/locales/en/translations.js [hunk 3]
```diff
@@ -76,7 +76,7 @@ export default {
     },
 
     'flexberry-maplayer': {
-      'opacity': 'Opacity'
+      'opacity': 'Visibility'
```
Comment (kuzkok) [inline addon/locales/en/translations.js]: why? may be need to rename property?

File: addon/layers/-private/base.js (hunk 2 above)
Comment (almostcake) [reply addon/layers/-private/base.js]: to correctly bind it in markup, like wgs84bbox.[i].[j]. Look [there](https://github.com/Flexberry/ember-flexberry-gis/blob/feature-144298-layer-extent/addon/templates/components/layers-dialogs/edit.hbs#L167)

File: addon/components/layers-dialogs/settings/wfs.js (hunk 1 above)
Comment (almostcake) [reply addon/components/layers-dialogs/settings/wfs.js]: it is inherited from WmsSettingsComponent where it is overriden and empty

File: addon/locales/en/translations.js (hunk 3 above)
Comment (almostcake) [reply addon/locales/en/translations.js]: I do not know why it is called opacity in leaflet, but I think that better to name properties according to their usage, and just to change caption. It is only caption

File: addon/components/base-layer.js [hunk 4]
```diff
@@ -466,6 +486,27 @@ export default Ember.Component.extend(
...
+      let layer = this.get('_leafletObject');
+
+      if (Ember.isNone(layer)) {
+        return new Ember.RSVP.Promise((resolve, reject) => {
+          reject(`Leaflet layer for '${this.get('layerModel.name')}' isn't created yet`);
+        });
+      }
+
+      let bounds = this._getBoundingBox(layer);
+      this.set('bounds', bounds);
```
Comment (kuzkok) [inline addon/components/base-layer.js]: is here right place for changing data, should we pass this bounds to model?

File: addon/components/base-layer.js [hunk 5]
```diff
@@ -466,6 +486,27 @@ export default Ember.Component.extend(
...
+      Returns leaflet layer's bounding box.
+
+      @method getBoundingBox
+      @return <a href="http://leafletjs.com/reference-1.1.0.html#latlngbounds">L.LatLngBounds</a>
+    */
+    getBoundingBox() {
+      let layer = this.get('_leafletObject');
+
+      if (Ember.isNone(layer)) {
+        return new Ember.RSVP.Promise((resolve, reject) => {
```
Comment (kuzkok) [inline addon/components/base-layer.js]: for which pusposes here returns promise?

File: addon/components/layers-dialogs/edit.js [hunk 6]
```diff
@@ -504,6 +562,51 @@ let FlexberryEditLayerDialogComponent = Ember.Component.extend(
...
+
+        // Coordinates should be projected to LatLngs.
+        if (boundsMode === 'bbox') {
+          let bbox = Ember.get(layer, 'settings.bbox');
+
+          if (!Ember.isBlank(bbox[0][0]) && !Ember.isBlank(bbox[0][1]) &&
+            !Ember.isBlank(bbox[1][0]) && !Ember.isBlank(bbox[1][1])) {
+
+            // Put it in property to force LeafletCRSMixin do it's work.
+            this.set('coordinateReferenceSystem', coordinateReferenceSystem);
```
Comment (kuzkok) [inline addon/components/layers-dialogs/edit.js]: property is unnecessary here.
you should create and export function in crs mixin and use it here.

File: addon/components/layers-dialogs/edit.js [hunk 7]
```diff
@@ -594,6 +697,27 @@ let FlexberryEditLayerDialogComponent = Ember.Component.extend(
...
+        this.set('boundsMode', newBoundsMode);
+      },
+
+      /**
+        Handles coordinate input textboxes keyPress events.
+
+        @method actions.coordsInputKeyPress
+      */
+      coordsInputKeyPress(e) {
+        if (e.which !== 45 && e.which !== 44 && e.which !== 46 && (e.which < 48 || e.which > 57)) {
```
Comment (kuzkok) [inline addon/components/layers-dialogs/edit.js]: what means that constants? should be in comment.

File: addon/mixins/flexberry-maplayer-actions-handler.js [hunk 8]
```diff
@@ -178,6 +177,60 @@ export default Ember.Mixin.create({
...
+      controllers/my-form.js
+      ```javascript
+        import Ember from 'ember';
+        import FlexberryMaplayerActionsHandlerMixin from 'ember-flexberry-gis/mixins/flexberry-maplayers-actions-handler';
+
+        export default Ember.Controller.extend(FlexberryMaplayerActionsHandlerMixin, {
+        });
+      ```
+    */
+    onMapLayerFitBounds(...args) {
```
Comment (kuzkok) [inline addon/mixins/flexberry-maplayer-actions-handler.js]:
```
onMapLayerFitBounds([boundsPropertyPath]) {
```
will be more readable

File: addon/models/new-platform-flexberry-g-i-s-map-layer.js [hunk 9]
```diff
@@ -144,6 +144,41 @@ let Model = Projection.Model.extend(NewPlatformFlexberyGISMapLayerModelMixin, Le
...
 
     return Ember.A();
+  }),
+
+  /**
+    Layer's latLngBounds.
+
+    @property bounds
+    @type <a href="http://leafletjs.com/reference-1.1.0.html#latlngbounds">L.LatLngBounds</a>
+    @return {Array} collection of child layers
```
Comment (kuzkok) [inline addon/models/new-platform-flexberry-g-i-s-map-layer.js]: return comment is wrong

File: addon/components/layers-dialogs/settings/wfs.js (hunk 1 above)
Comment (kuzkok) [reply addon/components/layers-dialogs/settings/wfs.js]: ok

File: addon/layers/-private/base.js (hunk 2 above)
Comment (kuzkok) [reply addon/layers/-private/base.js]: ok

File: addon/locales/en/translations.js (hunk 3 above)
Comment (kuzkok) [reply addon/locales/en/translations.js]: ok

File: addon/mixins/flexberry-maplayer-actions-handler.js (hunk 8 above)
Comment (almostcake) [reply addon/mixins/flexberry-maplayer-actions-handler.js]: ...args allow to put additional params anywhere by developer needs

File: addon/components/base-layer.js (hunk 5 above)
Comment (almostcake) [reply addon/components/base-layer.js]: Cuz private _getBoundingBox method returns a promise containing bbox, so the whole method should return a promise.

Comment (kuzkok) [reply addon/components/base-layer.js]: ok

File: addon/mixins/flexberry-maplayer-actions-handler.js (hunk 8 above)
Comment (kuzkok) [reply addon/mixins/flexberry-maplayer-actions-handler.js]: here you use only one param from args, why not do 
```
 let boundsPropertyPath = args[0];
``` 
already in method declaration

Comment (almostcake) [reply addon/mixins/flexberry-maplayer-actions-handler.js]: I think that following the whole structure is more important than slightly less complicated code

**Pull request merged**